### DISCUSSION TITLE
Support of multiple models in theia frontend and backend

### DISCUSSION THREAD

**Discussion by mishaltripathi** [eclipse-emfcloud/emfcloud#160] [ModelServer] [answered]
Hello All,

We are migrating our existing emf models to emf cloud in Theia application. However, there are couple of below open questions with respective to implementation.

**Background:**
Referring to very useful coffee editor example for model server where it has a single model server which runs on default port 8081 and using default model server client.

**Requirement:**
Additionally I need to create another feature which does have another model(let’s say Employee model) apart from coffee model.

**Solution:**
I am able to implement the same with below solution.
At backend: I have implemented another modelserver product which is very similar to existing model server and configured with 8082 port number.
At frontend: I have implemented custom modelserver client extension which is very similar to existing default model server client(i.e. DefaultModelServerClient) and configured with 8082 port number.

**Open questions:**
Can one model server backend product handle multiple models?
Is there any better to support multiple model servers apart from creating a multiple products running on different ports?
In case of multiple model server products, is there any restriction or recommendation on using specific port number for other model servers apart from 8081.
Any inputs on above is highly appreciated. Thanks in advance!!

**Comment by eneufeld** [comment, marked as answer]
Hi @mishaltripathi ,
Yes you can definitely handle multiple models in one backend.
Especially if you know all your models upfront and have no dynamic use case it is easy to.
You can find an example here:
https://github.com/eclipsesource/uml-glsp/blob/master/server/com.eclipsesource.uml.modelserver/src/main/java/com/eclipsesource/uml/modelserver/UmlModelServerModule.java

For dynamic cases this is more complex as there are some bindings that allow to be bound only once at the moment.

Best,
Eugen

**Reply by mishaltripathi** [reply]
Thanks @eneufeld for providing for your inputs. I will check this.
